fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java

```diff
@@ -299,12 +303,41 @@ public Cost visitPhysicalHashJoin(
                     0
             );
         }
-        return CostV1.of(context.getSessionVariable(), leftRowCount + rightRowCount + outputRowCount,
+        return CostV1.of(context.getSessionVariable(),
+                (leftRowCount + rightRowCount + outputRowCount)
+                        * skewPenalty(physicalHashJoin, probeStats, buildStats, totalInstanceNumber),
                 rightRowCount,
                 0
         );
     }
 
+    private double skewPenalty(PhysicalHashJoin<? extends Plan, ? extends Plan> physicalHashJoin,
+                               Statistics probeStats,
+                               Statistics buildStats,
+                               int totalInstanceNumber) {
+        double penalty = 1.0;
+        for (Expression conj : physicalHashJoin.getHashJoinConjuncts()) {
+            EqualTo eq = (EqualTo) JoinUtils.swapEqualToForChildrenOrder(
+                    (EqualTo) conj, physicalHashJoin.left().getOutputSet());
+            ColumnStatistic leftColStats = probeStats.findColumnStatistics(eq.left());
+            if (leftColStats != null && !leftColStats.isUnKnown()) {
+                if (leftColStats.ndv < beNumber) {
+                    penalty = totalInstanceNumber / Math.max(1.0, leftColStats.ndv);
```

Comment:
consider the most commonly used case, 3 be and leftColStats is 2 (meet < beNumber), the totalInstanceNumber is 24, then the penalty will be 12 on the original total cost for shuffle join, which seems too big changes...

Reply:
1. optimizer does not know the number of instances assigned to this PlanNode. totalInstanceNumber is the upper bound.
2. it is more important to balance workload between BEs than that of between instances, that is why penalty is triggered when "ndv < beNumber"